Add MangaHen

KenjieDec opened the pull request:
Closes #3805

Checklist:
- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [ ] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [x] Have removed `web_hi_res_512.png` when adding a new extension

KenjieDec committed:
Add MangaHen

KenjieDec committed:
Fix Title and Author

KenjieDec committed:
space

AwkwardPeak7 commented on src/en/mangahen/src/eu/kanade/tachiyomi/extension/en/mangahen/MangaHen.kt:
```diff
@@ -0,0 +1,172 @@
+
+        val mangas = doc.select("a[href^=/manga/]").map(::popularMangaFromElement)
+
+        val hasNextPage = doc.select("a[href*=page]").any { it.text().isEmpty() || it.text().isBlank() }
```
isBlank should be fine alone

AwkwardPeak7 commented on src/en/mangahen/src/eu/kanade/tachiyomi/extension/en/mangahen/MangaHen.kt:
```diff
@@ -0,0 +1,172 @@
+        val tagsList = response.asJsoup().select("li[onclick=updateTag(this)]").map { it.ownText().lowercase() }
+
+        return tagsList
+    }
```
fetch tag list once and cache them in a variable. currently you seem to be fetching them everytime in the loop for parsing search filters

or you can populat the filter list by fetching the tags too

AwkwardPeak7 commented on src/en/mangahen/src/eu/kanade/tachiyomi/extension/en/mangahen/MangaHen.kt:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    private fun tagsList(): List<String> {
+        val request = Request.Builder()
```
use GET or POST instead of  Request.Builder where possible

AwkwardPeak7 commented on src/en/mangahen/src/eu/kanade/tachiyomi/extension/en/mangahen/MangaHen.kt:
```diff
@@ -0,0 +1,172 @@
+                setUrlWithoutDomain(response.request.url.toString())
+            },
+        )
+    }
```
override fetchChapterList and return the chapter, since you are not using the response here

KenjieDec committed:
Fixed Chapter Images Quality, Apply Suggestions

- Use real images instead of thumbnail/preview images
- Apply AwkwardPeak's suggestions

AwkwardPeak7 commented on src/en/mangahen/MangaHen.kt:
you added the file in wrong place, move and merge it in the package directory (src/eu/kanade...)

KenjieDec committed:
Wrong Location1

KenjieDec committed:
Wrong Location2

KenjieDec committed:
Wrong Location3

AwkwardPeak7 commented on src/en/mangahen/src/eu/kanade/tachiyomi/extension/en/mangahen/MangaHen.kt:
```diff
@@ -0,0 +1,170 @@
+        val includeTags = mutableListOf<String>()
+        val excludeTags = mutableListOf<String>()
+
+        val tagsList = tagsList()
```
getting tags everytime isn't ideal, cache them in a variable.

also you can get them and populate FilterList dynamically, so users will not need to write them in text filter. [example implementation where genres are dynamically fetched ](https://github.com/keiyoushi/extensions-source/blob/main/src/all/galaxy/src/eu/kanade/tachiyomi/extension/all/galaxy/Galaxy.kt#L94-L157)

KenjieDec replied:
Thank you for the suggestion! I've changed the code according to your suggestion, is it correct?
\+ There are many tags (around 600). I have considered putting them in a select filter, but it would be inconvenient for users to scroll through such a long list just to find certain tags

KenjieDec committed:
Apply suggestion

- Apply AwkwardPeak's suggestion

KenjieDec committed:
curly bracket

KenjieDec committed:
space
